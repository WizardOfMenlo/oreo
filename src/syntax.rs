pub struct Identifier<'a>(&'a str);

pub struct Program<'a> {
    id: Identifier<'a>,
    compund: Compound<'a>,
}

pub struct Compound<'a> {
    statements: Vec<Statement<'a>>,
}

pub enum Statement<'a> {
    Decl(Decl<'a>),
    Print(Print<'a>),
    While(While<'a>),
    If(If<'a>),
    Assign(Assign<'a>),
}

pub struct Decl<'a> {
    id: Identifier<'a>,
    expr: Option<Expr<'a>>,
}

pub enum Print<'a> {
    Print(Expr<'a>),
    Println(Expr<'a>),
    Get(Identifier<'a>),
}

pub struct While<'a> {
    condition: Bool<'a>,
    compound: Compound<'a>,
}

pub struct If<'a> {
    condition: Bool<'a>,
    if_branch: Compound<'a>,
    else_branch: Option<Compound<'a>>,
}

pub struct Assign<'a> {
    id: Identifier<'a>,
    expr: Expr<'a>,
}

pub enum Expr<'a> {
    Addition(Box<BinaryOperator<'a>>),
    Subtraction(Box<BinaryOperator<'a>>),
    Multiplication(Box<BinaryOperator<'a>>),
    Division(Box<BinaryOperator<'a>>),
    Bool(Box<Bool<'a>>),
    Expr(Box<Expr<'a>>),
    Literal(Literal<'a>),
    Identifier(Identifier<'a>)
}

pub enum Literal<'a> {
    Str(&'a str),
    Int(isize)
}

pub struct BinaryOperator<'a> {
    left: Expr<'a>,
    right: Expr<'a>,
}

pub enum Bool<'a> {
    LessThan(BooleanRelOperator<'a>),
    GreaterThan(BooleanRelOperator<'a>),
    LesserOrEquals(BooleanRelOperator<'a>),
    GreaterOrEquals(BooleanRelOperator<'a>),
    Equals(BooleanRelOperator<'a>),
    And(Box<BooleanBinaryOperator<'a>>),
    Or(Box<BooleanBinaryOperator<'a>>),
    Not(Box<Bool<'a>>),
    BooleanLiteral(bool),
}

pub struct BooleanRelOperator<'a> {
    first: Expr<'a>,
    second: Expr<'a>,
}

pub struct BooleanBinaryOperator<'a> {
    first: Bool<'a>,
    second: Bool<'a>,
}