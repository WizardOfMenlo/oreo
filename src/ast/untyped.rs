//! The type we use to represent syntax nodes

use super::syntax::{AdditiveOp, BooleanOp, MultiplicativeOp, RelationalOp};
use crate::parser::error::SyntaxError;
use crate::range::Ranged;
use serde::Serialize;
use std::ops::Range;

/// We use this to figure out which kind of node we are
#[allow(missing_docs)]
#[derive(Debug, Clone, Serialize, PartialEq)]
pub enum NodeType<'a> {
    // Error node
    Error(SyntaxError<'a>),

    // Top
    Program,
    Compound,

    // Statements
    Decl,
    PrintStat,
    While,
    If,
    Assign,
    FunctionCall,
    FunctionDecl,
    Return,
    // Print vars
    Print,
    Println,
    Get,

    // Args
    FunctionDeclArgs,
    FunctionCallArgs,

    // Expr
    Expr,
    And,
    Or,

    // Term
    Term,
    LesserThan,
    GreaterThan,
    Equals,
    GreaterOrEquals,
    LesserOrEquals,

    // Factor
    Factor,
    Plus,
    Minus,

    // Product
    Product,
    Times,
    Divide,

    // Atom
    Unit,
    Not,

    // Unit
    Int(isize),
    Str,
    Bool(bool),
    Identifier,
    BracketedExpr,
}

impl<'a> NodeType<'a> {
    /// Is this node an error
    pub fn is_error(&self) -> bool {
        match self {
            NodeType::Error(_) => true,
            _ => false,
        }
    }

    /// If it is an error, get it, else panic
    pub fn unwrap_err(&self) -> SyntaxError {
        match self {
            NodeType::Error(e) => e.clone(),
            _ => panic!("Invalid conversion"),
        }
    }
}

/// The type that tracks type and range
#[derive(Debug, Clone, Serialize)]
pub struct Node<'a> {
    pub(super) ty: NodeType<'a>,
    pub(super) text_range: Range<usize>,
    pub(super) children: Vec<Node<'a>>,
}

impl<'a> Node<'a> {
    /// Creates  a new node
    pub fn new(ty: NodeType<'a>, range: Range<usize>, children: Vec<Node<'a>>) -> Self {
        Node {
            ty,
            text_range: range,
            children,
        }
    }

    /// Get the node type
    pub fn ty(&self) -> &NodeType<'a> {
        &self.ty
    }

    /// Get the node's children
    pub fn children(&self) -> impl DoubleEndedIterator<Item = &Node<'a>> {
        self.children.iter()
    }
}

impl<'a> Ranged for Node<'a> {
    fn range(&self) -> &Range<usize> {
        &self.text_range
    }
}

impl<'a> From<NodeType<'a>> for AdditiveOp {
    fn from(ty: NodeType) -> Self {
        match ty {
            NodeType::Plus => AdditiveOp::Plus,
            NodeType::Minus => AdditiveOp::Minus,
            _ => panic!("Invalid node type"),
        }
    }
}

impl<'a> From<NodeType<'a>> for MultiplicativeOp {
    fn from(ty: NodeType) -> Self {
        match ty {
            NodeType::Times => MultiplicativeOp::Times,
            NodeType::Divide => MultiplicativeOp::Divide,
            _ => panic!("Invalid node type"),
        }
    }
}

impl<'a> From<NodeType<'a>> for BooleanOp {
    fn from(ty: NodeType) -> Self {
        match ty {
            NodeType::And => BooleanOp::And,
            NodeType::Or => BooleanOp::Or,
            _ => panic!("Invalid node type"),
        }
    }
}

impl<'a> From<NodeType<'a>> for RelationalOp {
    fn from(ty: NodeType) -> Self {
        match ty {
            NodeType::LesserThan => RelationalOp::LesserThan,
            NodeType::GreaterThan => RelationalOp::GreaterThan,
            NodeType::LesserOrEquals => RelationalOp::LesserOrEquals,
            NodeType::GreaterOrEquals => RelationalOp::GreaterOrEquals,
            NodeType::Equals => RelationalOp::Equals,
            _ => panic!("Invalid node type"),
        }
    }
}
