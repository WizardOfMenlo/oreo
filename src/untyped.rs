use std::ops::Range;

use crate::parser::{ExpToken, TokenList};
use crate::range::RangedObject;
use crate::syntax::{AdditiveOp, BooleanOp, MultiplicativeOp, RelationalOp};
use crate::tokens::Token;

#[derive(Debug, Clone)]
pub enum SyntaxError<'a> {
    ExpectedOneOfFoundEOF(TokenList),
    ExpectedFoundEOF(ExpToken),
    ExpectedOneOfFound(TokenList, RangedObject<Token<'a>>),
    ExpectedFound(ExpToken, RangedObject<Token<'a>>),
    LogicalError,
}

#[derive(Debug, Clone)]
pub enum NodeType<'a> {
    // Error node
    Error(SyntaxError<'a>),

    // Top
    Program,
    Compound,

    // Statements
    Decl,
    PrintStat,
    While,
    If,
    Assign,
    FunctionCall,
    FunctionDecl,
    Return,
    // Print vars
    Print,
    Println,
    Get,

    // Args
    FunctionDeclArgs,
    FunctionCallArgs,

    // Expr
    Expr,
    And,
    Or,

    // Term
    Term,
    LesserThan,
    GreaterThan,
    Equals,
    GreaterOrEquals,
    LesserOrEquals,

    // Factor
    Factor,
    Plus,
    Minus,

    // Product
    Product,
    Times,
    Divide,

    // Atom (TODO: might be wrong)
    Unit,
    Not,

    // Unit
    Int(isize),
    Str,
    Bool(bool),
    Identifier,
    BracketedExpr,
}

impl<'a> NodeType<'a> {
    pub fn is_error(&self) -> bool {
        match self {
            NodeType::Error(_) => true,
            _ => false,
        }
    }
}

#[derive(Debug, Clone)]
pub struct Node<'a> {
    pub ty: NodeType<'a>,
    pub text_range: Range<usize>,
    pub children: Vec<Node<'a>>,
}

impl<'a> From<NodeType<'a>> for AdditiveOp {
    fn from(ty: NodeType) -> Self {
        match ty {
            NodeType::Plus => AdditiveOp::Plus,
            NodeType::Minus => AdditiveOp::Minus,
            _ => panic!("Invalid node type"),
        }
    }
}

impl<'a> From<NodeType<'a>> for MultiplicativeOp {
    fn from(ty: NodeType) -> Self {
        match ty {
            NodeType::Times => MultiplicativeOp::Times,
            NodeType::Divide => MultiplicativeOp::Divide,
            _ => panic!("Invalid node type"),
        }
    }
}

impl<'a> From<NodeType<'a>> for BooleanOp {
    fn from(ty: NodeType) -> Self {
        match ty {
            NodeType::And => BooleanOp::And,
            NodeType::Or => BooleanOp::Or,
            _ => panic!("Invalid node type"),
        }
    }
}

impl<'a> From<NodeType<'a>> for RelationalOp {
    fn from(ty: NodeType) -> Self {
        match ty {
            NodeType::LesserThan => RelationalOp::LesserThan,
            NodeType::GreaterThan => RelationalOp::GreaterThan,
            NodeType::LesserOrEquals => RelationalOp::LesserOrEquals,
            NodeType::GreaterOrEquals => RelationalOp::GreaterOrEquals,
            NodeType::Equals => RelationalOp::Equals,
            _ => panic!("Invalid node type"),
        }
    }
}
