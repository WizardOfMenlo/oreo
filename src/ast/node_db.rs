//! Something

use super::untyped::Node;
use std::collections::HashMap;
use std::marker::PhantomPinned;
use std::pin::Pin;
use std::ptr::NonNull;

/// An id for a node
#[derive(Debug, Clone, Copy, Hash, Eq, PartialEq, Ord, PartialOrd)]
pub struct NodeId(usize);

/// A node that is flat
#[derive(Debug, Clone)]
pub struct IdNode {
    childrens: Vec<NodeId>,
}

impl IdNode {
    /// Gets the children of this node
    pub fn children(&self) -> &[NodeId] {
        &self.childrens
    }
}

/// A database that efficiently allows for querying and retrivieng nodes
#[derive(Debug)]
pub struct NodeDb<'a> {
    node: Node<'a>,
    id_nodes: HashMap<NodeId, IdNode>,
    nodes: HashMap<NodeId, NonNull<Node<'a>>>,
    start_node_id: NodeId,
    _phantom: PhantomPinned,
}

impl<'a> NodeDb<'a> {
    /// Creates a db from a node
    pub fn new(node: Node<'a>) -> Pin<Box<Self>> {
        // Ok this is complex and merit some discussion
        // In order to efficiently (aka not storing nodes and children twice)
        // store nodes, we aim to store a reference to the children of a stored not
        // Ordinarly, rust wouldn't like that, so we need to bend the rules a little
        // First of all, this is a self referential struct, i.e. one that has references upon itself
        // (in particular node is borrowed by nodes)
        // Since moving the struct would invalidate this references, we use the Pin API, which allows to
        // to define types that cannot be moved
        // Using this, we can make our thing work and be efficient!
        let db = NodeDb {
            node,
            _phantom: PhantomPinned,
            // Temp
            id_nodes: HashMap::new(),
            nodes: HashMap::new(),
            start_node_id: NodeId(0),
        };

        let mut boxed = Box::pin(db);

        let (id, id_nodes, nodes) =
            add_node(&boxed.node, NodeId(0), HashMap::new(), HashMap::new());

        // I claim this is safe, as we are never moving the box and such not
        // violating the Pin API
        unsafe {
            let reference = Pin::as_mut(&mut boxed);
            let inner = Pin::get_unchecked_mut(reference);
            inner.start_node_id = NodeId(id.0.saturating_sub(1));
            inner.id_nodes = id_nodes;
            inner.nodes = nodes;
        }

        boxed
    }

    /// All the ids that this struct owns
    pub fn get_all_node_ids<'b>(&'b self) -> impl Iterator<Item = NodeId> + 'b {
        self.id_nodes.keys().cloned()
    }

    /// Get a flat representation of the node
    pub fn get_flat_node(&self, id: NodeId) -> Option<&IdNode> {
        self.id_nodes.get(&id)
    }

    /// Get the start id (ie the top node id)
    pub fn start_id(&self) -> NodeId {
        self.start_node_id
    }

    /// Get a reference to the full node
    pub fn get_node(&self, id: NodeId) -> Option<&Node<'a>> {
        // Note this should be safe as p always refer to a node in the
        // struct
        self.nodes.get(&id).map(|p| unsafe { p.as_ref() })
    }

    /// Get all the children that this node has
    pub fn all_children(&self, id: NodeId) -> Vec<NodeId> {
        let mut res = Vec::new();
        for child in self
            .get_flat_node(id)
            .map(|n| n.children())
            .unwrap_or_default()
        {
            res.push(*child);
            res.extend(self.all_children(*child));
        }
        res
    }
}

type NodeIdMap = HashMap<NodeId, IdNode>;
type NodeMap<'a> = HashMap<NodeId, NonNull<Node<'a>>>;

// Not in class to circumvent (safely) rust borrowing rules
fn add_node<'a>(
    n: &Node<'a>,
    mut id: NodeId,
    mut id_nodes: NodeIdMap,
    mut nodes: NodeMap<'a>,
) -> (NodeId, NodeIdMap, NodeMap<'a>) {
    let mut ids = Vec::new();
    for c in n.children() {
        let (n_id, n_id_nodes, n_nodes) = add_node(c, id, id_nodes.clone(), nodes.clone());
        id = n_id;
        id_nodes.extend(n_id_nodes);
        nodes.extend(n_nodes);
        ids.push(NodeId(id.0 - 1));
    }

    id_nodes.insert(id, IdNode { childrens: ids });

    nodes.insert(id, NonNull::from(n));

    (NodeId(id.0 + 1), id_nodes, nodes)
}

#[cfg(test)]
mod tests {
    use super::*;
    use insta::assert_debug_snapshot;

    fn node_from_str(input: &str) -> Node {
        use crate::lexer::lexicalize;
        use crate::lexer::scanner::scan;
        use crate::parser::parse;

        parse(lexicalize(scan(input)))
    }

    // We use this to test, as the Db does not serialize consistently
    fn get_all_nodes<'a>(db: &NodeDb<'a>) -> Vec<Node<'a>> {
        let mut ids: Vec<_> = db.get_all_node_ids().collect();
        ids.sort_by(|a, b| a.0.cmp(&b.0));

        ids.into_iter()
            .map(|i| db.get_node(i).expect("This should be failproof"))
            .cloned()
            .collect()
    }

    #[test]
    fn node_db_minimal() {
        let input = "program fib begin var n; end";
        let db = NodeDb::new(node_from_str(input));
        let nodes = get_all_nodes(&db);
        assert_debug_snapshot!(nodes);
    }

    #[test]
    fn node_db_minimal_assignments() {
        let input = "program fib begin var n := \"Hello\"; var m := 2; var s := (2); end";
        let db = NodeDb::new(node_from_str(input));
        let nodes = get_all_nodes(&db);
        assert_debug_snapshot!(nodes);
    }

    #[test]
    fn node_db_print_simple() {
        let input = "program fib begin print \"Hello\\n\"; println 2; get x; end";
        let db = NodeDb::new(node_from_str(input));
        let nodes = get_all_nodes(&db);
        assert_debug_snapshot!(nodes);
    }
}
