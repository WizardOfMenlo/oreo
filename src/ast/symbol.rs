//! Utilities for figuring out symbols and scopes

use super::node_db::NodeDb;
use super::syntax::*;
use std::collections::HashMap;

/// An id for an identifier
#[derive(Debug, Clone, Copy, Hash, Eq, PartialEq)]
pub struct IdentId(usize);

/// A id for a scope
#[derive(Debug, Clone, Copy, Hash, Eq, PartialEq)]
pub struct ScopeId(usize);

/// A full symbol table
#[derive(Debug, Default)]
pub struct SymbolTable<'a> {
    scopes: HashMap<ScopeId, Scope<'a>>,
}

/// A scope
#[derive(Debug, Clone)]
pub struct Scope<'a> {
    parent: ScopeId,
    variables: PartialSymbolTable<'a>,
}

/// A symbol table for a scope
#[derive(Debug, Default, Clone)]
pub struct PartialSymbolTable<'a> {
    vars: Vec<VariableRecord<'a>>,
}

/// A possible declaration
#[derive(Debug, Clone)]
pub enum DeclarationContext {
    /// var x := 0
    Decl(Decl),
    /// procedure f(var f) ...
    FunctionArg(FunctionDeclArgs),
}

/// A record for a variable (i.e. declaration and name)
#[derive(Debug, Clone)]
pub struct VariableRecord<'a> {
    id: IdentId,
    text: &'a str,
    decl: DeclarationContext,
}

/// Build the symbol table from a node
#[derive(Debug)]
pub struct SymbolTableBuilder<'a, 'b> {
    current_id: IdentId,
    current_scope: ScopeId,
    // Used to generate new ids
    scope_count: usize,
    symb: SymbolTable<'a>,
    input: &'a str,
    db: &'b NodeDb<'a>,
}

impl<'a, 'b> SymbolTableBuilder<'a, 'b> {
    /// Starts building the symbol table
    pub fn new(input: &'a str, db: &'b NodeDb<'a>) -> Self {
        SymbolTableBuilder {
            current_id: IdentId(0),
            current_scope: ScopeId(0),
            scope_count: 0,
            symb: SymbolTable::default(),
            input,
            db,
        }
    }

    /// Build the full table
    pub fn build(mut self, program: Program) -> SymbolTable<'a> {
        self.compound(program.compound(self.db));

        self.symb
    }

    fn get_current_scope(&mut self) -> &mut Scope<'a> {
        self.symb
            .scopes
            .get_mut(&self.current_scope)
            .expect("Invalid scope")
    }

    fn enter_new_scope(&mut self) {
        dbg!(&self.symb);
        let new_scope = Scope {
            parent: self.current_scope,
            variables: PartialSymbolTable::default(),
        };

        self.scope_count += 1;
        self.current_scope = ScopeId(self.scope_count);
        self.symb.scopes.insert(self.current_scope, new_scope);

        dbg!(&self.symb);
    }

    fn exit_scope(&mut self) {
        self.current_scope = self.get_current_scope().parent;
    }

    fn add_var(&mut self, id: Identifier, decl: DeclarationContext) {
        let curr_id = self.current_id;
        let id_str = id.id(self.db, self.input);
        self.get_current_scope()
            .variables
            .vars
            .push(VariableRecord {
                id: curr_id,
                text: id_str,
                decl,
            });

        self.current_id = IdentId(self.current_id.0 + 1);
    }

    fn compound(&mut self, compound: Compound) {
        self.enter_new_scope();
        for statement in compound.statements(self.db) {
            match statement.downcast(self.db) {
                StatementType::Decl(d) => {
                    self.add_var(d.id(self.db), DeclarationContext::Decl(d));
                }
                StatementType::FunctionDecl(f) => {
                    self.enter_new_scope();
                    for arg in f.args(self.db) {
                        self.add_var(arg.id(self.db), DeclarationContext::FunctionArg(arg));
                    }
                    self.compound(f.compound(self.db));
                    self.exit_scope();
                }
                StatementType::If(i) => {
                    self.compound(i.if_branch(self.db));
                    if let Some(c) = i.else_branch(self.db) {
                        self.compound(c);
                    }
                }
                StatementType::While(w) => {
                    self.compound(w.compound(self.db));
                }
                _ => continue,
            }
        }
        self.exit_scope();
    }
}

#[cfg(test)]
mod tests {
    use super::*;
    use insta::assert_debug_snapshot;
    use std::pin::Pin;

    fn db_from_str(input: &str) -> Pin<Box<NodeDb>> {
        use crate::lexer::lexicalize;
        use crate::lexer::scanner::scan;
        use crate::parser::parse;

        let node = parse(lexicalize(scan(input)));
        NodeDb::new(node)
    }

    // We use this to have a determined order
    fn determinize<'a>(table: SymbolTable<'a>) -> Vec<(ScopeId, Scope<'a>)> {
        let mut v: Vec<_> = table.scopes.into_iter().collect();
        v.sort_by(|a, b| (a.0).0.cmp(&(b.0).0));
        v
    }

    #[test]
    fn test_simple() {
        let input = "program x begin var x := 10; end";
        let db = db_from_str(input);
        let table = SymbolTableBuilder::new(input, &db).build(Program::new(db.start_id()));
        assert_debug_snapshot!(determinize(table));
    }

    #[test]
    fn test_nested() {
        let input = r#"program id begin
                   if ( true ) then begin var n := 5;  end
       			   else begin var k := 46; end; end"#;
        let db = db_from_str(input);
        let table = SymbolTableBuilder::new(input, &db).build(Program::new(db.start_id()));
        assert_debug_snapshot!(determinize(table));
    }

    #[test]
    fn test_fun_decl() {
        let input = r#"
        program id begin
            procedure f(var x, var y) begin
                var z;
                var k;
            end
       	end"#;
        let db = db_from_str(input);
        let table = SymbolTableBuilder::new(input, &db).build(Program::new(db.start_id()));
        assert_debug_snapshot!(determinize(table));
    }
}
