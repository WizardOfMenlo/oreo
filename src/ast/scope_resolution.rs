//! Module that does variable resolution

use super::node_db::NodeDb;
use super::symbol::{IdentId, ScopeId, SymbolTable, GLOBAL_SCOPE};
use super::syntax::*;
use std::collections::HashMap;

/// Struct that stores the mapping between identifier and actual vars
#[derive(Debug)]
pub struct VariableResolver {
    id_mapping: HashMap<Identifier, IdentId>,
}

/// We use this to build the resolver
#[derive(Debug)]
pub struct VariableResolverBuilder<'a, 'b, 'c> {
    current_scope: ScopeId,
    parents: Vec<ScopeId>,
    scope_counter: usize,
    input: &'a str,
    symbols: &'b SymbolTable<'a>,
    db: &'c NodeDb<'a>,
}

/// An error in resolution
#[derive(Debug)]
pub struct ResolutionError<'a> {
    string: &'a str,
    scope: ScopeId,
    identifier: Identifier,
}

impl<'a, 'b, 'c> VariableResolverBuilder<'a, 'b, 'c> {
    /// Initializes the builder
    pub fn new(input: &'a str, symbols: &'b SymbolTable<'a>, node_db: &'c NodeDb<'a>) -> Self {
        VariableResolverBuilder {
            current_scope: GLOBAL_SCOPE,
            input,
            symbols,
            scope_counter: 0,
            parents: Vec::new(),
            db: node_db,
        }
    }

    /// Build from the program
    pub fn build(mut self, program: Program) -> Result<VariableResolver, Vec<ResolutionError<'a>>> {
        let mut res = HashMap::new();
        let mut errs = Vec::new();

        self.compound(program.compound(self.db), &mut res, &mut errs);

        if errs.is_empty() {
            Ok(VariableResolver { id_mapping: res })
        } else {
            Err(errs)
        }
    }

    fn enter_scope(&mut self) {
        self.parents.push(self.current_scope);
        self.scope_counter += 1;
        self.current_scope = self.scope_counter.into();
    }

    fn exit_scope(&mut self) {
        self.current_scope = self.parents.pop().expect("Tried to exit global scope");
    }

    fn resolve_id(
        &mut self,
        id: Identifier,
        map: &mut HashMap<Identifier, IdentId>,
        errs: &mut Vec<ResolutionError<'a>>,
    ) {
        let parents = self.symbols.parent_scopes(self.current_scope);
        let id_str = id.id(self.db, self.input);

        let mut found = false;
        for scope in parents {
            if let Some(id_id) = self.symbols.get_id_scope(id_str, scope) {
                found = true;
                map.insert(id, id_id);
                break;
            }
        }

        if !found {
            errs.push(ResolutionError {
                string: id_str,
                scope: self.current_scope,
                identifier: id,
            });
        }
    }

    fn expr(
        &mut self,
        expr: Expr,
        map: &mut HashMap<Identifier, IdentId>,
        errs: &mut Vec<ResolutionError<'a>>,
    ) {
        use super::untyped::NodeType;
        for children in self.db.all_children(expr.get_id()) {
            if let NodeType::Identifier = self.db.get_node(children).expect("Invalid node id").ty()
            {
                self.resolve_id(Identifier::new(children), map, errs);
            }
        }
    }

    fn compound(
        &mut self,
        compound: Compound,
        map: &mut HashMap<Identifier, IdentId>,
        errs: &mut Vec<ResolutionError<'a>>,
    ) {
        self.enter_scope();

        for statement in compound.statements(self.db) {
            match statement.downcast(self.db) {
                StatementType::Decl(d) => {
                    self.resolve_id(d.id(self.db), map, errs);
                    if let Some(e) = d.expr(self.db) {
                        self.expr(e, map, errs);
                    }
                }
                StatementType::Assign(a) => {
                    self.resolve_id(a.id(self.db), map, errs);
                    self.expr(a.expr(self.db), map, errs);
                }
                StatementType::FunctionCall(f) => {
                    self.resolve_id(f.id(self.db), map, errs);
                    for arg in f.args(self.db) {
                        self.expr(arg.expr(self.db), map, errs)
                    }
                }
                StatementType::FunctionDecl(f) => {
                    self.resolve_id(f.id(self.db), map, errs);
                    self.enter_scope();
                    for arg in f.args(self.db) {
                        self.resolve_id(arg.id(self.db), map, errs);
                    }
                    self.compound(f.compound(self.db), map, errs);
                    self.exit_scope();
                }
                StatementType::If(i) => {
                    self.expr(i.condition(self.db), map, errs);
                    self.compound(i.if_branch(self.db), map, errs);
                    if let Some(c) = i.else_branch(self.db) {
                        self.compound(c, map, errs);
                    }
                }
                StatementType::While(w) => {
                    self.expr(w.condition(self.db), map, errs);
                    self.compound(w.compound(self.db), map, errs);
                }
                StatementType::Return(r) => self.expr(r.expr(self.db), map, errs),
                StatementType::PrintStat(p) => match p.downcast(self.db) {
                    PrintTypes::Get(g) => self.resolve_id(g.id(self.db), map, errs),
                    PrintTypes::Print(p) => self.expr(p.expr(self.db), map, errs),
                    PrintTypes::Println(p) => self.expr(p.expr(self.db), map, errs),
                },
            }
        }

        self.exit_scope();
    }
}

#[cfg(test)]
mod tests {
    use super::super::node_db::NodeDbWrap;
    use super::super::symbol::{SymbolTable, SymbolTableBuilder, VariableRecord};
    use super::*;
    use insta::assert_debug_snapshot;

    fn db_from_str(input: &str) -> NodeDbWrap {
        use crate::lexer::lexicalize;
        use crate::lexer::scanner::scan;
        use crate::parser::parse;

        let node = parse(lexicalize(scan(input)));
        NodeDb::new(node)
    }

    fn sym_table_from<'a>(input: &'a str, db: &NodeDb<'a>) -> SymbolTable<'a> {
        SymbolTableBuilder::new(input, db).build(Program::new(db.start_id()))
    }

    use super::super::untyped::Node;

    // We use this to have a determined order
    fn determinize<'a>(
        resolver: VariableResolver,
        db: &NodeDb<'a>,
        sym: &'a SymbolTable<'a>,
    ) -> Vec<(Node<'a>, Identifier, VariableRecord<'a>)> {
        let mut res = Vec::new();
        for (id_n, ident_id) in resolver.id_mapping {
            let node = id_n.get_node(&db).clone();
            let record = sym.get_variable_record(ident_id).unwrap();
            res.push((node, id_n, record))
        }

        // We want to fix the order, else all will break randomly
        res.sort_by(|a, b| {
            a.2.id()
                .cmp(&b.2.id())
                .then(a.1.get_id().cmp(&b.1.get_id()))
        });

        res
    }

    #[test]
    fn scope_res_simple() {
        let input = "program x begin var x := 10; var y := x + 1; end";
        let db = db_from_str(input);
        let sym = sym_table_from(input, &db);
        let resolver =
            VariableResolverBuilder::new(input, &sym, &db).build(Program::new(db.start_id()));
        assert_debug_snapshot!(resolver.map(|r| determinize(r, &db, &sym)));
    }

    #[test]
    fn scope_res_nested() {
        let input = r#"
        program id begin
            var top := 10;
            if ( true ) then
            begin
                var n := top;
                var l := n;
            end
            else 
            begin 
                var k := top; 
            end;
        end"#;
        let db = db_from_str(input);
        let sym = sym_table_from(input, &db);
        let resolver =
            VariableResolverBuilder::new(input, &sym, &db).build(Program::new(db.start_id()));
        assert_debug_snapshot!(resolver.map(|r| determinize(r, &db, &sym)));
    }

    #[test]
    fn scope_res_nested_invalid() {
        let input = r#"
        program id begin
            if ( true ) then
            begin
                var n := 1;
            end
            else 
            begin 
                var k := n; 
            end;
        end"#;
        let db = db_from_str(input);
        let sym = sym_table_from(input, &db);
        let resolver =
            VariableResolverBuilder::new(input, &sym, &db).build(Program::new(db.start_id()));
        assert_debug_snapshot!(resolver.map(|r| determinize(r, &db, &sym)));
    }

    #[test]
    fn scope_res_test_fun_decl() {
        let input = r#"
        program id begin
            procedure f(var x ~ bool, var y ~ int) begin
                var z;
                var k := z;
            end
       	end"#;
        let db = db_from_str(input);
        let sym = sym_table_from(input, &db);
        let resolver =
            VariableResolverBuilder::new(input, &sym, &db).build(Program::new(db.start_id()));
        assert_debug_snapshot!(resolver.map(|r| determinize(r, &db, &sym)));
    }
}
