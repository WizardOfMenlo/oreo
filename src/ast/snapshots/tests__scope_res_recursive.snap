---
source: src/ast/scope_resolution.rs
expression: "resolver.map(|r| determinize(r, &db, &sym))"
---
Ok(
    [
        (
            Node {
                ty: Identifier,
                text_range: 40..41,
                children: [],
            },
            Identifier(
                NodeId(
                    1,
                ),
            ),
            VariableRecord {
                id: IdentId(
                    0,
                ),
                text: "y",
                decl: Decl(
                    Decl(
                        NodeId(
                            8,
                        ),
                    ),
                ),
            },
        ),
        (
            Node {
                ty: Identifier,
                text_range: 74..75,
                children: [],
            },
            Identifier(
                NodeId(
                    10,
                ),
            ),
            VariableRecord {
                id: IdentId(
                    1,
                ),
                text: "f",
                decl: FunctionDecl(
                    FunctionDecl(
                        NodeId(
                            20,
                        ),
                    ),
                ),
            },
        ),
        (
            Node {
                ty: Identifier,
                text_range: 119..120,
                children: [],
            },
            Identifier(
                NodeId(
                    11,
                ),
            ),
            VariableRecord {
                id: IdentId(
                    1,
                ),
                text: "f",
                decl: FunctionDecl(
                    FunctionDecl(
                        NodeId(
                            20,
                        ),
                    ),
                ),
            },
        ),
    ],
)
